### ISSUE TITLE
bug: RestoreObject Fails with Internal Error (500) on Non-existent Keys

### ISSUE THREAD

**Issue by anna-mazhar** [localstack/localstack#10332]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

- When invoking the restore_object boto SDK operation on a non-existent key in LocalStack, the operation fails with a 500 HTTP code. 
- The error message reported is an "Internal Error" stating: "'NoneType' object has no attribute 'storage_class'".
- Exception handling seems to be missing.

### Expected Behavior

- It should fail with a 404 HTTP status code and the error message "The specified key does not exist."

### How are you starting LocalStack?

With a `docker run` command

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

    docker run localstack/localstack

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

    s3_client = boto3.client('s3', endpoint_url='http://localhost:4566')
    response = s3_client.restore_object(
        Bucket='test_bucket',
        Key='nonexistent-key'
    )

The object key does not exist.

### Environment

```markdown
- OS: Ubuntu 22.04.2 LTS
- LocalStack: 2.1.0
- boto3: 1.28.6
```

**Comment by bentsku** [comment]
Hello @anna-mazhar and thanks for your report! Like #10331, it seems this issue has been fixed in the meantime in versions > 3.

I could not reproduce:

```python
response = s3_client.restore_object(
    Bucket=bucket,
    Key='nonexistent-key'
)
> botocore.errorfactory.NoSuchKey: An error occurred (NoSuchKey) when calling the RestoreObject operation (reached max retries: 0): The specified key does not exist.
```

Could you please confirm that it now works? Thanks!

**Comment by anna-mazhar** [comment]
Hi @bentsku, thanks for getting back! I confirm that this has been fixed in 3.2.0.
